lib/DxilPIXPasses/PixPassHelpers.cpp

```diff
@@ -37,17 +37,49 @@ using namespace llvm;
 using namespace hlsl;
 
 namespace PIXPassHelpers {
-bool IsAllocateRayQueryInstruction(llvm::Value const *Val) {
+static bool
+IsRayQueryHandleRecursive(llvm::Value const *Val,
+                          std::vector<llvm::Value const *> &PreviousPhis) {
   if (Val != nullptr) {
     if (llvm::Instruction const *Inst =
             llvm::dyn_cast<llvm::Instruction>(Val)) {
-      return hlsl::OP::IsDxilOpFuncCallInst(Inst,
-                                            hlsl::OP::OpCode::AllocateRayQuery);
+      if (hlsl::OP::IsDxilOpFuncCallInst(Inst,
+                                         hlsl::OP::OpCode::AllocateRayQuery)) {
+        return true;
+      } else if (auto PHI = llvm::dyn_cast<llvm::PHINode>(Inst)) {
+        for (unsigned IVOrdinal = 0; IVOrdinal < PHI->getNumIncomingValues();
+             ++IVOrdinal) {
+          // phi instructions can recursively refer to each other
+          if (std::find(PreviousPhis.begin(), PreviousPhis.end(), Val) ==
```

Comment:
This is really gross, so I hope somebody has a better idea about how to cope with circular phi references...

Reply:
Can't you keep a set of visited instructions and pass a reference to that set to IsRayQueryHandleRecursive?  Skip the incoming value if it's in the set.

Reply:
You could use a `llvm::SmallPtrSet`, and you could just try adding it when you have a PHI, returning false if it was already in the set.  Like:
```
if (!VisitedPhis.insert(PHI).second)
  return false;
```

Reply:
Maybe I missed your point, but that seems equivalent to what I have? And if I don't clean up the set as I go back up the stack, the set could grow to many thousands (not really a "small" pointer set) and the lookup could bog down.

Reply:
Another approach is to just make a map of PHI instructions that are RayQuery handles by following the AllocateRayQuery users.  I think this might be a lot better approach.  This should end up being a small map or empty map most of the time, and you just look up the PHI in the map to see if it is one.

Reply:
Good idea. Worked well.